### PR TITLE
samples: bluetooth: mesh_provisioner: push button to provision

### PR DESCRIPTION
**Pull request by HaavardRei**
In order to keep a flashed sample from provisioning forever, the sample
requires the user to press a button to provision a detected device.

The feature is only enabled if the board has a push button connected
via GPIO.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

### CODE REVIEW

**Comment by Andrewpini** [comment]
@jhedberg can you approve the workflow for this PR? This is a task from the Nordic Mesh team backlog that is done by one of our summer interns.

File: samples/bluetooth/mesh_provisioner/src/main.c
Comment (Andrewpini) [inline samples/bluetooth/mesh_provisioner/src/main.c:310]: Might sound a little bit strange, but in general we try to use `k_work_delayable` rather than timers. Could you try to duplicate the behaviour using `k_work_delayable`?

File: samples/bluetooth/mesh_provisioner/src/main.c
Comment (Andrewpini) [inline samples/bluetooth/mesh_provisioner/src/main.c:357]: This could be wraped in a separate `button_init` function

Comment (Andrewpini) [inline samples/bluetooth/mesh_provisioner/src/main.c:357]:
```suggestion
	printk("Press button 1 to enable the Provisioner.\n);
```

**Review by Andrewpini — COMMENTED** [review]
I have tested this and it seems to be working fine 👍 I have a couple of suggestions so please take a look at my comments.

**Review by Andrewpini — COMMENTED** [review]
Looks good to me, but have you reviewed the README for the sample? Does it need to be updated after the added changes?

File: samples/bluetooth/mesh_provisioner/src/main.c
Comment (Andrewpini) [inline samples/bluetooth/mesh_provisioner/src/main.c:316]:
```suggestion
static void button_init(void)
```

File: samples/bluetooth/mesh_provisioner/src/main.c
Comment (Andrewpini) [inline samples/bluetooth/mesh_provisioner/src/main.c:310]:
```suggestion
static void button_pressed(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
```

**Comment by PavelVPV** [comment]
Wouldn't it be better to provision the found device only after button press?

**Comment by Andrewpini** [comment]
> Wouldn't it be better to provision the found device only after button press?

Well, that would work to. I don't have a strong opinion on the approach as long as we get rid of the provision forever feature. If your up for it @HaavardRei you can try to do what @PavelVPV suggested.

**Comment by HaavardRei** [comment]
@PavelVPV , @Andrewpini I reworked the code using a semaphore to synchronize provisioning with a button press. Should squash the previous commits before merging if this is approved.

File: samples/bluetooth/mesh_provisioner/src/main.c
Comment (alxelax) [inline samples/bluetooth/mesh_provisioner/src/main.c:15]: What about platforms `**qemu_cortex_m3 qemu_x86**`? They are in the allowed platforms list.
Is it still possible to build and run app for them?

Comment (alxelax) [reply samples/bluetooth/mesh_provisioner/src/main.c:15]: I'm not sure about the importance of platforms. This is a question mostly to the code owner.
However, I think if it is not possible to build an application for the board it shouldn't be in the list of allowed platforms.

Comment (PavelVPV) [reply samples/bluetooth/mesh_provisioner/src/main.c:15]: (not a code owner of this sample) No, it won't build. But a user may want to run this sample in QEMU or as native application and use BlueZ as a controller. This variant is also mentioned in the documentation. As a suggestion, it can either be compiled with button option for boards that has a button and without button option for other boards (like QEMU).

Comment (HaavardRei) [reply samples/bluetooth/mesh_provisioner/src/main.c]: I've done as @PavelVPV suggested and in the latest commit the feature is only enabled if the board supports GPIO. The sample builds for all specified platforms.

File: samples/bluetooth/mesh_provisioner/src/main.c
Comment (PavelVPV) [inline samples/bluetooth/mesh_provisioner/src/main.c:352]: Minor: If I'm not mistaken, preprocessor directives should not have indentation.

File: samples/bluetooth/mesh_provisioner/src/main.c
Comment (PavelVPV) [inline samples/bluetooth/mesh_provisioner/src/main.c:372]: Minor: 3 minutes can be too much. May be 30 seconds?

**Review by PavelVPV — CHANGES_REQUESTED** [review]
Few last minor comments from my side.

You can now update the PR description and title according to the current changes.
Also, feel free to force push commits, github will show delta to the previous changes.

**Review by Andrewpini — APPROVED** [review]
LGTM 👍

**Review by PavelVPV — DISMISSED** [review]

**Review by Andrewpini — DISMISSED** [review]

**Comment by Andrewpini** [comment]
I see that compliance is complaining about a couple of long lines in the documentation. Please take a look and resolve it.

**Review by alxelax — DISMISSED** [review]

**Review by Andrewpini — DISMISSED** [review]

**Comment by PavelVPV** [comment]
@HaavardRei you need to shorten lines in commit message to 75 symbols.

**Comment by HaavardRei** [comment]
@PavelVPV , @alxelax It should be good now :)

**Review by PavelVPV — APPROVED** [review]

**Review by Andrewpini — APPROVED** [review]

**Review by alxelax — APPROVED** [review]

**Pull request merged**
